lib/oplog/interval_max_metric_test.go

```diff
@@ -0,0 +1,307 @@
+			m.Report(12)
+			m.Report(13)
+			m.Report(14)
+			m.Report(2)
+			m.Report(0)
+			m.Report(-29)
+			m.Report(12)
+
+			m.Collect(c)
+			req.Len(c, 0)
+		})
+
+		m.opts.withNow(now.Add(interval), func() {
+			m.Collect(c)
+			req.Len(c, 1)
+
+			asrt.Equal(14., val(t, <-c))
+
+		})
+
+		m.opts.withNow(now.Add(2*interval), func() {
+			m.Collect(c)
+			req.Len(c, 0)
+
+			m.Report(52)
+			m.Report(-12)
+			m.Report(0)
+			m.Report(512395)
+			m.Report(18)
+		})
+
+		m.opts.withNow(now.Add(3*interval), func() {
+			m.Collect(c)
+			req.Len(c, 1)
+
+			asrt.Equal(512395., val(t, <-c))
+
+			m.Report(0)
+		})
+
+		m.opts.withNow(now.Add(4*interval), func() {
+			m.Collect(c)
+			req.Len(c, 1)
+			asrt.Equal(0., val(t, <-c))
+
+			m.Report(-1)
+		})
+
+		m.opts.withNow(now.Add(5*interval), func() {
+			m.Collect(c)
+			req.Len(c, 1)
+			asrt.Equal(-1., val(t, <-c))
+		})
+	})
+}
+
+func val(t *testing.T, metric prometheus.Metric) float64 {
+	d := dto.Metric{}
+	require.NoError(t, metric.Write(&d))
+
+	return *d.Gauge.Value
+}
+
+func labels(t *testing.T, metric prometheus.Metric) map[string]string {
+	d := dto.Metric{}
+	require.NoError(t, metric.Write(&d))
+
+	ret := map[string]string{}
+
+	for _, pair := range d.Label {
+		ret[*pair.Name] = *pair.Value
+	}
+
+	return ret
+}
+
+func TestIntervalMaxMetricVec(t *testing.T) {
+	t.Parallel()
+
+	const interval = 5 * time.Millisecond
+
+	req := require.New(t)
+	asrt := assert.New(t)
+
+	m := NewIntervalMaxMetricVec(&IntervalMaxVecOpts{
+		IntervalMaxOpts: IntervalMaxOpts{
+			Opts:           prometheus.Opts{},
+			ReportInterval: interval,
+		},
+		GCInterval: interval,
+	}, []string{"l1", "l2"})
+
+	now := time.Now()
+
+	c := make(chan prometheus.Metric, 16)
+	m.opts.withNow(now, func() {
+		m.Report(12, "a", "test")
+		m.Report(0, "another", "test")
+		m.Report(13, "a", "test")
+		m.Report(-1, "another", "test")
+
+		m.Collect(c)
+		req.Len(c, 0)
+	})
+
+	m.opts.withNow(now.Add(interval), func() {
+		m.Collect(c)
+		req.Len(c, 2)
+
+		for i := 0; i < 2; i++ {
+			v := <-c
+
+			l := labels(t, v)
+			value := val(t, v)
+
+			if l["l1"] == "a" {
+				asrt.Equal(13., value)
+			} else {
+				asrt.Equal(0., value)
+			}
+		}
+	})
+
+	m.opts.withNow(now.Add(3*interval), func() {
+		// ensure a gc happens
+		count := 0
+		m.mp.Range(func(_, _ interface{}) bool {
+			count++
+			return true
+		})
+		asrt.Equal(2, count)
+
+		m.Collect(c)
+
+		count = 0
+		m.mp.Range(func(_, _ interface{}) bool {
+			count++
+			return true
+		})
+		asrt.Equal(0, count)
+	})
+}
+
+func TestIntervalMaxMetric_TimeTravelingPanic(t *testing.T) {
+	t.Parallel()
+
+	const interval = 5 * time.Millisecond
+
+	asrt := assert.New(t)
+
+	m := NewIntervalMaxMetric(&IntervalMaxOpts{
+		Opts:           prometheus.Opts{},
+		ReportInterval: interval,
+	}, []string{"l1", "l2"}, []string{"a", "b"})
+
+	now := time.Now()
+
+	m.opts.withNow(now.Add(5*interval), func() {
+		m.Report(4.0)
+	})
+
+	asrt.Panics(func() {
```

Comment:
Isn't the idea of using the monotonic clock that we *avoid* this panic?

Reply:
Yes&mdash;the only time I can think that this would happen is if the monotonic clock overflows. Based on what I've read, we shouldn't expect to ever see this happen, even though technically the POSIX spec permits `CLOCK_MONOTONIC` to be set up in such a way that it overflows, say, 10 minutes after boot:

>For this clock, the value returned by `clock_gettime()` represents the amount of time (in seconds and nanoseconds) since an **unspecified point in the past** (for example, system start-up time, or the Epoch).

Realistically, we shouldn't ever see the behavior, but I figured since it was possible, I might as well handle it and test it. I'll note this in the other comment.